### PR TITLE
More accurate shape of optimization

### PR DESCRIPTION
**Pull request by jane-intel**
### Details:
 - *Topological order was corrected to perform the best ShapeOf reconnections possible*

### Tickets:
 - *CVS-142953*

### CODE REVIEW

File: src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp [hunk 1]
```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
...
+        }
+    }
+    // step 2: starting from Result -- assign weight to ops:
+    //      if parameter dependant, weights is maximum of output indices plus one
+    //      else weights is maximum of output indices
+    // this step doesn't assign weights to all the ops, this is intentional and will be used in the following step
+    const std::string weight_rt_info_name = "topological_sort_weight";
+    auto reverse_order = m->get_ordered_ops();
+    std::reverse(reverse_order.begin(), reverse_order.end());
+    for (const auto& op : reverse_order) {
```
Comment (praasz) [inline src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:248]: Maybe is possible to not use `std::reverse` but replace `for` loop using reverse iterator from model's ordered ops?

File: src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp [hunk 2]
```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
...
+    // finalization: descending order for levels and ops within level are ordered by get_ordered_ops
+    std::vector<std::shared_ptr<ov::Node>> result;
+    result.reserve(reverse_order.size());
+    for (const auto& item : level_to_vector)
+        result.insert(result.end(), item.second.rbegin(), item.second.rend());
+    for (const auto& op : result) {
+        op->get_rt_info().erase(weight_rt_info_name);
+        op->get_rt_info().erase(op_depends_on_parameter);
+    }
+    std::reverse(result.begin(), result.end());
```
Comment (praasz) [inline src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:312]: Is possible to insert into `result` in reverse order then using reverse would not be required?

File: src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp [hunk 3]
```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
...
+        if (ov::as_type_ptr<ov::op::v0::Parameter>(op)) {
+            op->get_rt_info()[op_depends_on_parameter] = true;
+        } else if (ov::as_type_ptr<ov::op::v0::Constant>(op) || ov::as_type_ptr<ov::op::v0::ShapeOf>(op) ||
+                   ov::as_type_ptr<ov::op::v3::ShapeOf>(op) ||
+                   std::dynamic_pointer_cast<ov::op::util::VariableExtension>(op)) {
+            op->get_rt_info()[op_depends_on_parameter] = false;
+        } else {  // deduce op type from inputs
+            const auto& inputs = op->input_values();
+            op->get_rt_info()[op_depends_on_parameter] =
+                std::any_of(inputs.begin(), inputs.end(), [=](const ov::Output<ov::Node>& input) {
```
Comment (praasz) [inline src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:236]:
```suggestion
                std::any_of(inputs.begin(), inputs.end(), [&op_depends_on_parameter](const ov::Output<ov::Node>& input) {
```
`=` the `op_depends_on_parameter` will be copied, is required to do it.

File: src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp [hunk 4]
```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
...
+    // step 3: make propagation for all the nodes:
+    // if weight is already assigned -- skip operation
+    // else operation weights is minimum of input indices
+    // if all operation inputs have no weights -- this op is isolated and this algorithm doesn't make sense,
+    // such cases are extremely rare and rather theoretical, to handle them we return original ov::Model op order
+    std::map<int64_t, std::vector<std::shared_ptr<ov::Node>>> level_to_vector;
+    for (const auto& op : m->get_ordered_ops()) {
+        if (op->get_rt_info().count(weight_rt_info_name)) {
+            level_to_vector[op->get_rt_info().at(weight_rt_info_name).as<int64_t>()].push_back(op);
             continue;
```
Comment (praasz) [inline src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:282]: Can be removed as two code block excluding

File: src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp [hunk 5]
```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
...
+            int64_t weight = 0;
+            bool inputs_have_weight = false;
+            for (const auto& input : op->input_values()) {
+                const auto& rt_info = input.get_node_shared_ptr()->get_rt_info();
+                if (!rt_info.count(weight_rt_info_name))
+                    continue;
+                auto input_weight = rt_info.at(weight_rt_info_name).as<int64_t>();
+                if (!inputs_have_weight)
+                    weight = input_weight;
+                weight = input_weight < weight ? input_weight : weight;
```
Comment (praasz) [inline src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:293]:
```suggestion
                if (input_weight < weight)
                    weight = input_weight;
```
Set `weight` to i64 max and on first check will be always get `input_weight`, on next step will change only if `input_weight` is lower then current `weight`

**Review by praasz — APPROVED** [review]

File: src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp [hunk 6]
```diff
@@ -218,7 +219,101 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
...
 
-void save_shape_sources(const ov::Output<ov::Node>& output, STS_map& symbol_shape_source) {
+std::vector<std::shared_ptr<ov::Node>> topological_order(const std::shared_ptr<ov::Model>& m) {
+    auto order = m->get_ordered_ops();
+
+    // step 1: split model into parameter related and parameter non-related ops
+    const std::string op_depends_on_parameter = "topological_sort_op_depends_on";
+    // values: true - parameter dependent; false otherwise
+    for (const auto& op : order) {
+        if (ov::as_type_ptr<ov::op::v0::Parameter>(op)) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:229]: should we also track ReadValue ops?

Comment (itikhono) [reply src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:229]: I found in the 2nd If branch
std::dynamic_pointer_cast<ov::op::util::VariableExtension>(op)) {

Comment (jane-intel) [reply src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:229]: No, we are only interested in the parameter dependant operations.
In steps 1 and 2 we are extracting the "main" graph -- the longest path from Parameter to Result, other graph paths will be ordered according to their connection to the "main" graph.

Please take a look at the [presentation](https://intel-my.sharepoint.com/:p:/p/evgeniia_nugmanova/EddCf8lCm9lFj87R2Xm0duQBZTJ4rb75tuB19G6cxWDTrw?e=jbNQFr) for a better understanding

Comment (itikhono) [reply src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp:229]: ok, I will, thanks!

**Review by WeldonWangwang — APPROVED** [review]

**Comment by peterchen-intel** [comment]
@jane-intel Is it possible to have a test case to make sure this design will not be changed unconsciously?

**Comment by jane-intel** [comment]
@peterchen-intel there is a test that has been changed. It will fail if the logic changes in the future.

**Pull request merged**